File: apps/api/src/routes/strategy.ts
import type { FastifyInstance } from 'fastify';
import { strategySyncService } from '../services/strategy-sync.js';
import type { StrategyCategory, StrategySubcategory } from '../types.js';
import { configRoutes } from './config.js';
import { previewRoutes } from './preview.js';

/**
 * 判断策略是否"工作流就绪"：
 * category ∈ canonical 3 值集 AND subcategory ∈ canonical 10 值集 → true
 * subcategory=null → false（未分类策略视为未就绪）
 */
const CANONICAL_CATEGORIES: StrategyCategory[] = ['factor_based', 'non_factor', 'transitional'];
const CANONICAL_SUBCATEGORIES: StrategySubcategory[] = [
  'linear_multi_factor', 'index_enhancement', 'ml_nonlinear_factor',
  'trend_cta', 'arbitrage', 'hft_microstructure', 'macro_quant',
  'event_driven', 'e2e_ai_timeseries', 'event_sentiment_factor',
];

function isWorkflowReady(category: string, subcategory: string | null): boolean {
  if (subcategory === null) return false;
  return (
    CANONICAL_CATEGORIES.includes(category as StrategyCategory) &&
    CANONICAL_SUBCATEGORIES.includes(subcategory as StrategySubcategory)
  );
}

function mapMeta(m: Awaited<ReturnType<typeof strategySyncService.syncFromPython>>[number]) {
  return {
    name: m.name,
    description: m.description,
    params: m.params.map((p) => ({
      key: p.key,
      label: p.label,
      type: p.type,
      default: p.default,
      min: p.min,
      max: p.max,
      options: p.options,
      chart_relevant: p.chart_relevant ?? false,
      ui_constraints: p.ui_constraints ?? [],
    })),
    version: m.version,
    kind: m.kind,
    backtestable: m.backtestable,
    category: m.category ?? 'non_factor',
    subcategory: m.subcategory ?? null,
    workflowReady: isWorkflowReady(m.category ?? 'non_factor', m.subcategory ?? null),
  };
}

export async function strategyRoutes(app: FastifyInstance) {
  // ── 策略元信息 ──

  app.get('/', async () => {
    const strategies = await strategySyncService.syncFromPython();
    return strategies.map(mapMeta);
  });

  app.get<{ Params: { name: string } }>('/:name', async (req, reply) => {
    const strategies = await strategySyncService.syncFromPython();
    const meta = strategies.find((m) => m.name === req.params.name);
    if (!meta) return reply.code(404).send({ error: 'Strategy not found' });
    return {
      ...mapMeta(meta),
      modes: meta.modes,
    };
  });

  // ── 策略配置 CRUD（挂载 GET/PUT /:name/config） ──
  await app.register(configRoutes);

  // ── Preview 端点（挂载 POST /:name/preview） ──
  await app.register(previewRoutes);
}

File: packages/obsidian-sync/quantforge_obsidian/builders/strategy.py
"""策略笔记构建器"""

from __future__ import annotations

from quantforge_strategy import StrategyMeta


def build_strategy_overview(strategies: list[StrategyMeta]) -> str:
    items = "\n".join(
        f"- **{s.name}** — {s.description}（模式: {', '.join(m.value for m in s.modes)}）→ [[策略/{s.name}]]"
        for s in strategies
    )
    return f"""---
tags: [quant/strategy, moc]
---

# 策略概览

共 {len(strategies)} 个策略。

{items}

---

*由 QuantForge obsidian-sync 自动生成*"""


def build_strategy_note(meta: StrategyMeta) -> str:
    if meta.params:
        rows = []
        for p in meta.params:
            rng = f"[{p.min}, {p.max}]" if p.type.value == "number" and p.min is not None and p.max is not None else "-"
            rows.append(f"| {p.key} | {p.type.value} | {p.default} | {rng} |")
        params_table = "| 参数 | 类型 | 默认值 | 范围 |\n|------|------|--------|------|\n" + "\n".join(rows)
    else:
        params_table = "无"

    modes = ", ".join(m.value for m in meta.modes)

    required_factors = ""
    if meta.required_factors:
        links = "\n".join(f"- [[因子/{f}]]" for f in meta.required_factors)
        required_factors = f"\n\n## 依赖因子\n\n{links}"

    return f"""---
type: strategy
version: {meta.version}
modes: [{modes}]
tags: [quant/strategy]
---

# {meta.name}

## 描述

{meta.description}

## 参数

{params_table}

## 模式

{modes}{required_factors}

---

*由 QuantForge obsidian-sync 自动生成*"""
